"Error write empty file to Hive" while writing to bucketed hive table

While inserting a row into a bucketed hive table, the following error occurs:


```shell
$ java -jar /opt/presto-cli-317-executable.jar --server localhost:8080  --user user
presto>  DROP TABLE hive.myschema.page_views;
DROP TABLE
presto> CREATE TABLE hive.myschema.page_views (user_id bigint, page_url varchar, dt date) WITH ( partitioned_by = ARRAY['dt'], bucketed_by = ARRAY['user_id'], bucket_count = 10 );
CREATE TABLE
presto> insert into hive.myschema.page_views values (cast(123 as bigint), 'http://url.com', cast(date_parse('01/01/2016 6:06:05 PM','%m/%d/%Y %h:%i:%s %p') as date));

Query 20190814_213058_00002_jnn4q, FAILED, 1 node
Splits: 35 total, 34 done (97.14%)
1:29 [0 rows, 0B] [0 rows/s, 0B/s]

Query 20190814_213058_00002_jnn4q failed: Error write empty file to Hive
```

```java
io.prestosql.spi.PrestoException: Error write empty file to Hive
        at io.prestosql.plugin.hive.HiveMetadata.writeEmptyFile(HiveMetadata.java:1309)
        at io.prestosql.plugin.hive.HiveMetadata.createEmptyFile(HiveMetadata.java:1294)
        at io.prestosql.plugin.hive.HiveMetadata.finishInsert(HiveMetadata.java:1384)
        at io.prestosql.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.finishInsert(ClassLoaderSafeConnectorMetadata.java:394)
        at io.prestosql.metadata.MetadataManager.finishInsert(MetadataManager.java:778)
        at io.prestosql.sql.planner.LocalExecutionPlanner.lambda$createTableFinisher$3(LocalExecutionPlanner.java:2768)
        at io.prestosql.operator.TableFinishOperator.getOutput(TableFinishOperator.java:297)
        at io.prestosql.operator.Driver.processInternal(Driver.java:379)
        at io.prestosql.operator.Driver.lambda$processFor$8(Driver.java:283)
        at io.prestosql.operator.Driver.tryWithLock(Driver.java:675)
        at io.prestosql.operator.Driver.processFor(Driver.java:276)
        at io.prestosql.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:1075)
        at io.prestosql.execution.executor.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:163)
        at io.prestosql.execution.executor.TaskExecutor$TaskRunner.run(TaskExecutor.java:484)
        at io.prestosql.$gen.Presto_317____20190814_205908_1.run(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: DestHost:destPort host1:25000 , LocalHost:localPort host10/10.0.0.10:0. Failed on local exception: java.io.IOException: org.apache.hadoop.security.AccessControlException: Client cannot authe
nticate via:[TOKEN, KERBEROS]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.apache.hadoop.net.NetUtils.wrapWithMessage(NetUtils.java:831)
        at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:806)
        at org.apache.hadoop.ipc.Client.getRpcResponse(Client.java:1515)
        at org.apache.hadoop.ipc.Client.call(Client.java:1457)
        at org.apache.hadoop.ipc.Client.call(Client.java:1367)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:228)
        at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:116)
        at com.sun.proxy.$Proxy179.getFileInfo(Unknown Source)
        at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.getFileInfo(ClientNamenodeProtocolTranslatorPB.java:903)
        at sun.reflect.GeneratedMethodAccessor732.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:422)
        at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeMethod(RetryInvocationHandler.java:165)
        at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invoke(RetryInvocationHandler.java:157)
        at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeOnce(RetryInvocationHandler.java:95)
        at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:359)
        at com.sun.proxy.$Proxy180.getFileInfo(Unknown Source)
        at org.apache.hadoop.hdfs.DFSClient.getFileInfo(DFSClient.java:1665)
        at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1582)
        at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1579)
        at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
        at org.apache.hadoop.hdfs.DistributedFileSystem.getFileStatus(DistributedFileSystem.java:1594)
        at org.apache.hadoop.fs.FilterFileSystem.getFileStatus(FilterFileSystem.java:454)
        at org.apache.hadoop.fs.FileSystem.exists(FileSystem.java:1683)
        at org.apache.hadoop.fs.FileSystem.createNewFile(FileSystem.java:1359)
        at org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat$OrcRecordWriter.close(OrcOutputFormat.java:111)
        at io.prestosql.plugin.hive.HiveMetadata.writeEmptyFile(HiveMetadata.java:1306)
        ... 17 more
Caused by: java.io.IOException: org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
        at org.apache.hadoop.ipc.Client$Connection$1.run(Client.java:760)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:422)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
        at org.apache.hadoop.ipc.Client$Connection.handleSaslConnectionFailure(Client.java:723)
        at org.apache.hadoop.ipc.Client$Connection.setupIOstreams(Client.java:817)
        at org.apache.hadoop.ipc.Client$Connection.access$3700(Client.java:411)
        at org.apache.hadoop.ipc.Client.getConnection(Client.java:1572)
        at org.apache.hadoop.ipc.Client.call(Client.java:1403)
        ... 41 more
Caused by: org.apache.hadoop.security.AccessControlException: Client cannot authenticate via:[TOKEN, KERBEROS]
        at org.apache.hadoop.security.SaslRpcClient.selectSaslClient(SaslRpcClient.java:173)
        at org.apache.hadoop.security.SaslRpcClient.saslConnect(SaslRpcClient.java:390)
        at org.apache.hadoop.ipc.Client$Connection.setupSaslConnection(Client.java:617)
        at org.apache.hadoop.ipc.Client$Connection.access$2300(Client.java:411)
        at org.apache.hadoop.ipc.Client$Connection$2.run(Client.java:804)
        at org.apache.hadoop.ipc.Client$Connection$2.run(Client.java:800)
        at java.security.AccessController.doPrivileged(Native Method)
        at javax.security.auth.Subject.doAs(Subject.java:422)
        at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1730)
        at org.apache.hadoop.ipc.Client$Connection.setupIOstreams(Client.java:800)
        ... 44 more
```

The same query works fine for a non-bucketed table. The issue occurs when presto is creating the bucket files. I did some debugging and I found that presto first creates bucket files in the hdfs `/tmp` dir, and I can see the bucket file is present there.

```java
2019-08-15T05:41:13.293+0800    DEBUG   20190814_214110_00002_d3c4m.1.0-16-148  io.prestosql.plugin.hive.HiveEventClient File created: query: 20190814_214110_00002_d3c4m, schema: myschema, table: page_views, partition: 'dt=201
6-01-01', format: org.apache.hadoop.hive.ql.io.orc.OrcOutputFormat, size: 408, path: hdfs://hacluster/tmp/presto-user/e29a0898-7a76-4e06-9f9f-d678feac9c9f/dt=2016-01-01/000003_0_20190814_214110_00002_d3c4m
2019-08-15T05:41:13.388+0800    DEBUG   20190814_214110_00002_d3c4m.1.0-16-148  io.prestosql.execution.executor.TaskExecutor     Split 20190814_214110_00002_d3c4m.1.0-16  (start = 1.95273490613619E9, wall = 1234 ms, cpu = 411 m
s, wait = 0 ms, calls = 2) is finished
2019-08-15T05:41:13.424+0800    DEBUG   query-execution-1       io.prestosql.execution.StageStateMachine        Stage 20190814_214110_00002_d3c4m.1 is FINISHED
2019-08-15T05:41:13.530+0800    INFO    20190814_214110_00002_d3c4m.0.0-16-125  stdout  Java config name: /opt/presto/config/krb5.conf
2019-08-15T05:41:13.530+0800    INFO    20190814_214110_00002_d3c4m.0.0-16-125  stdout  Loaded from Java config
2019-08-15T05:41:13.530+0800    INFO    20190814_214110_00002_d3c4m.0.0-16-125  stdout  >>> KdcAccessibility: reset
2019-08-15T05:41:13.531+0800    INFO    20190814_214110_00002_d3c4m.0.0-16-125  stdout  >>>KinitOptions cache name is /tmp/krb5cc_0
2019-08-15T05:42:56.281+0800    ERROR   SplitRunner-11-125      io.prestosql.execution.executor.TaskExecutor    Error processing Split 20190814_214110_00002_d3c4m.0.0-16  (start = 1.952734903892547E9, wall = 104129 ms, cpu = 18
 ms, wait = 17 ms, calls = 4): HIVE_WRITER_CLOSE_ERROR: Error write empty file to Hive

```

```shell
$ hdfs dfs -ls /tmp/presto-user/e29a0898-7a76-4e06-9f9f-d678feac9c9f/dt=2016-01-01/000003_0_20190814_214110_00002_d3c4m
-rw-r--r--   3 user supergroup        408 2019-08-15 05:41 /tmp/presto-user/e29a0898-7a76-4e06-9f9f-d678feac9c9f/dt=2016-01-01/000003_0_20190814_214110_00002_d3c4m

```

According to the logs, it seems to fail while closing the bucket file it's writing. However, I don't see how it could be an authentication issue if presto was able to create the table and then the bucket file in the tmp dir using the same kerberos user in the same session.

The weird part is that I have tried deploying presto onto two machine and on one I run into the issue but not the other. I have the exact same configs under `/opt/presto-server-317/etc`. Using the exact same hive and kerberos config files. Also the same java version.

I also ran into the issue on a third machine but when I cleared the `/tmp` dir in hdfs, it started working. However, the same solution is not working on the failing machine.

Would anyone has any idea as to what might be causing this issue? What other differences should I look for between the working machine and the failing one.

I am using the official release presto-server-317.tar.gz and presto-cli-317-executable.jar files.

--- a/presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java
+++ b/presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java
@@ -1167,7 +1167,7 @@ public class HiveMetadata
             partitionUpdates = PartitionUpdate.mergePartitionUpdates(concat(partitionUpdates, partitionUpdatesForMissingBuckets));
             for (PartitionUpdate partitionUpdate : partitionUpdatesForMissingBuckets) {
                 Optional<Partition> partition = table.getPartitionColumns().isEmpty() ? Optional.empty() : Optional.of(buildPartitionObject(session, table, partitionUpdate));
-                createEmptyFile(session, partitionUpdate.getWritePath(), table, partition, partitionUpdate.getFileNames());
+                createEmptyFiles(session, partitionUpdate.getWritePath(), table, partition, partitionUpdate.getFileNames());
             }
         }
 
@@ -1275,7 +1275,7 @@ public class HiveMetadata
         return missingFileNames;
     }
 
-    private void createEmptyFile(ConnectorSession session, Path path, Table table, Optional<Partition> partition, List<String> fileNames)
+    private void createEmptyFiles(ConnectorSession session, Path path, Table table, Optional<Partition> partition, List<String> fileNames)
     {
         JobConf conf = toJobConf(hdfsEnvironment.getConfiguration(new HdfsContext(session, table.getDatabaseName(), table.getTableName()), path));
 
@@ -1289,10 +1289,11 @@ public class HiveMetadata
             schema = getHiveSchema(table);
             format = table.getStorage().getStorageFormat();
         }
-
-        for (String fileName : fileNames) {
-            writeEmptyFile(session, new Path(path, fileName), conf, schema, format.getSerDe(), format.getOutputFormat());
-        }
+        hdfsEnvironment.doAs(session.getUser(), () -> {
+            for (String fileName : fileNames) {
+                writeEmptyFile(session, new Path(path, fileName), conf, schema, format.getSerDe(), format.getOutputFormat());
+            }
+        });
     }
 
     private static void writeEmptyFile(ConnectorSession session, Path target, JobConf conf, Properties properties, String serDe, String outputFormatName)
@@ -1381,7 +1382,7 @@ public class HiveMetadata
             partitionUpdates = PartitionUpdate.mergePartitionUpdates(concat(partitionUpdates, partitionUpdatesForMissingBuckets));
             for (PartitionUpdate partitionUpdate : partitionUpdatesForMissingBuckets) {
                 Optional<Partition> partition = table.getPartitionColumns().isEmpty() ? Optional.empty() : Optional.of(buildPartitionObject(session, table, partitionUpdate));
-                createEmptyFile(session, partitionUpdate.getWritePath(), table, partition, partitionUpdate.getFileNames());
+                createEmptyFiles(session, partitionUpdate.getWritePath(), table, partition, partitionUpdate.getFileNames());
             }
         }
 

